### DISCUSSION TITLE
2007/08/08/labview-%E4%B8%AD%E7%9A%84%E6%B3%9B%E5%9E%8B%E5%AE%B9%E5%99%A8

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#176] [Announcements]
# 2007/08/08/labview-%E4%B8%AD%E7%9A%84%E6%B3%9B%E5%9E%8B%E5%AE%B9%E5%99%A8

https://lv.qizhen.xyz/oop\generic

https://blog.qizhen.xyz/2007/08/08/labview-%E4%B8%AD%E7%9A%84%E6%B3%9B%E5%9E%8B%E5%AE%B9%E5%99%A8

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Stephen Mercer LV R&D
2017年12月12日 03:33 编辑
Qizhen: Do you still have the xnode VIs that you created for this project? I have been looking everywhere for them and cannot find them. I am hoping to resurrect the effort to add this functionality to LabVIEW.

回复
ruanqizhen
2021年11月8日 06:01 编辑
Hi Stephen, Just saw your message. I haven’t logged in for so many years. I didn’t keep the code. This project was once saved in NI Labs, but I couldn’t find it any more. No sure if you can still find it internally.

回复
Stephen
2021年11月8日 09:37 编辑
Thanks for checking. I couldn’t find it internally either. That’s a shame. You did good work on it. With some of the new language features, I think I could really make these shine. Oh well.

回复
ruanqizhen
2021年11月8日 15:35 编辑
I’m so glad that you tried to integrate it into LabVIEW. I never touched LabVIEW again after leaving NI. I used lots of other languages but LabVIEW is still my favorite.

回复
Pingback： 博客版《我和LabVIEW》目录 « 凡人琐忆 编辑
噜噜
2011年04月6日 15:48 编辑
刚好想在labview中使用map，但是在空间中未能找到项目源码的下载链接。。现想实现从数据库中读出内容存入map中，这样就不用每次都去读数据库，减少时间开销。对该项目很感兴趣，能否传一份源码来参考下，谢谢~

回复
ruanqizhen
2011年04月7日 08:56 编辑
这个软件内部是使用LabVIEW代码实现了一份红黑树。它的版权不属于我个人，所以我不能把代码给你。
如果你的数据比较简单，有个更方便的方法满足你的要求，你可以看一下这篇文章：
https://ruanqizhen.wordpress.com/2010/08/11/%E5%A6%82%E6%9E%9C%E9%9C%80%E8%A6%81%E5%9C%A8labview%E4%B8%AD%E4%BD%BF%E7%94%A8%E5%88%B0%E5%AD%97%E5%85%B8%E5%AE%B9%E5%99%A8/

回复
Qizhen
2008年10月30日 04:58 编辑
XNode 尚未成熟，所以目前还没有对外开放编辑它的权限。

回复
彬
2008年10月29日 17:57 编辑
在你开发的Map中使用到LabVIEW 8.0后才出现的*.xnode，请问我们要怎么去编辑属于自己的*.xnode？

回复
lijuansophie
2008年07月17日 16:40 编辑
太强了，敬仰一下

回复
Qizhen
2007年12月30日 13:44 编辑
后来更新过，主要是修改了一些bug。已经可以生成EXE了。新的版本只能在8.5下使用。哈希表，我们考虑过，但是没有尝试实现，感觉上它比二叉树要难，主要没有一个各种情况都适用的哈希算法。对于某一个特定问题，倒是可以考虑做个哈希表的。

回复
Chien-Chi
2007年09月29日 22:26 编辑
我目前使用LabVIEW 8.5版本在測試發現問題和在之前LabVIEW 8.2版本一樣,就是當我在value type複製一個constant時將它更新名稱,以此例是複製"職工編號"將它重新名命為"Index",並且將它的order改為0,然而,在建立map out時卻出現下圖左的indicator和下圖右的cluster的constant不同,導致必須手動更改indicator的名稱,很不方便


回复
tata
2007年08月16日 15:27 编辑
 完全不懂，这是什么？
老大，你在过强悍了
弄的我只有惊叹的份
因为不懂
不过还是谢谢你的留言喽
呵呵
每一篇下面该有篇日志会好一点，估计会来的人多一点
人多有什么好处不用说你也明白的
宣传啊，同行就夹在其间，呵呵

回复
老虎
2007年08月15日 09:57 编辑
这个空间很牛，要顶，要常来。。

回复
Nicholas
2007年08月12日 11:10 编辑
好久没来了。一来就看见经典，呵呵，顶一个。虽然现在不用LV，未必以后也不用，还是把你的BLOG加到我的GOOGLE RSS Reader里吧。多发文呀～～没事也上俺“家”瞅瞅

回复
杰
2007年08月10日 17:43 编辑
谢谢回复。
不能编译成exe就很麻烦了。我只能自己使用，而无法发布我的产品。因为目标机器上不可能安装LV8.2。LV一个lisence还是挺贵的。不过能自己用用也挺满足了，毕竟现在还是free的。呵呵。用map还是能省不少事的，尤其是可以嵌套，这玩意好使。
我自己在用LV写和map的功能差不多的子vi，只是可能数据类型支持的不够多，算法不够好而已。其实map里面的很多功能我都写了，只是在看了你编写的map这样一个概念之后，知道了这样的原型，加深了对程序结构、数据处理流程的理解了。很有收获。我可以在自己的机器上编程的时候，就用map，发布exe的时候，就把map相应的东西换成我自己的vi。
曲线救国啊，呵呵。

回复
Qizhen
2007年08月10日 14:57 编辑
现在这个 Map 在内存占用上并不能比以前减少，它只是利用的平衡二叉树，把搜索的时间复杂度从 O(n)降低到了 O(log n)。它是提高程序运行速度，但不能减少程序的内存消耗。如果数据量特别特别大，可能运行起来还是有问题。这不是 Map 的问题，而是LabVIEW 在这方面存在固有缺陷引起的。这个我现在也没有想到很好的解决方法。

回复
Qizhen
2007年08月10日 14:46 编辑
容器里可以放数据，也可以放其它容器。所以 Map 是可以嵌套的。吧一个 Map 作为数据放到另一个 Map 中去。（今天上午我跟你讨论过这个问题，不过当时我好像说错了）

回复
Qizhen
2007年08月10日 14:37 编辑
secondary key 这样的参数以后是不会提供的，因为有些应用可能还需要第三、第四关键字等等，我们不能吧函数的参数越添越多。解决的方法有两个，一个是短期解决方案，你可以把你的第一关键字、第二关键字，组成一个 cluster。把新构成的 cluster 作为关键字，就可以了。长期，我有个计划，就是支持用户定义“比较”行为，你也提到这点了。对于两个同类型的数据，用户可以指定它们比较的方式，而不是像现在这样直接调用LabVIEW的大于号来比较。比如，可以定义先比较数据中的一个元素，相等，就再比较另一个元素，知道区分出大小。这样以来，用户就可以自己对数据排序了。

回复
Qizhen
2007年08月10日 14:27 编辑
manual 和 help 里面的错误我会马上更正的。

回复
Qizhen
2007年08月10日 14:26 编辑
这个软件应该可以在任何操作系统下的 LabVIEW 8.2 中使用，这是我只对 Windows XP 进行过测试。但是如果没有 LabVIEW 8.2 就不行了。我今天试了一下把它升成可执行文件，结果出了错。以前没试过做exe，所以没意识到这个错误。这个无法升成可执行文件的问题，会争取在下个版本解决。

回复
Qizhen
2007年08月10日 14:21 编辑
回答 GaoJie 的问题首先感谢你来用，像这样一个还在试验阶段的项目，能够由用户来用，就是对他的最大肯定了。

回复
杰
2007年08月9日 22:04 编辑
前面点了两下，发重复了。呵呵，删了一个吧。
刚刚用了一下你的这个东东，还是挺好使的。
能否提供一个secondary key?也就是说，你的sort key下，能不能定义下一层的key？这样应该会更容易用到复杂应用中。呵呵。其实也就是说，map是一个容器，他可以包含任何类型的数据，那么，能够包含另外一个map吗？ 
还有，你的这个实验品，基本上就是查询的功能，排序的好像没看见，是固定排序的吧？我其实非常需要compare 的功能。如果有整体key之间的比较和key包含的基本value的比较，那么会很好用。呵呵。
从使用方法上来看，使用map，我首先需要的是把数据都放到map里面去，然后才能查询。但其实如果我已经都获得了数据，我直接比较查询就能够得出结果了。编程也不复杂，使用map能明显提高效率，还是减小内存的消耗？个人觉得，如果数据都放在map里面，数据量少没有问题，如果多了，即使map只提供的是内存的指针（我不知道你是不是这样开发的），也会比较耗内存，我现在处理的方式是查一块之后，丢掉一块，否则根本不跑不动。
顺便说下，我download下来的manual里面好像有点小错误，你检查下，一个是erase的定义，写成了find，估计是从上面copy下来的没改，嘿嘿，俺也经常干这样的事情。还有一个，valid好像没定义。其实象这样的要是能直接加到LV的帮助里面就好了。现在好像content里面只有简单的提示，没有detail的帮助，是吧？
最后还要问下，OS如果是2000的，是不是没法用这个啊？单位里面都是2000的系统，只有家里才是xp。要不在xp底下打包成exe或者installer到单位用？

回复
杰
2007年08月9日 20:26 编辑
非常好，我刚刚用LV完成一个海量测试数据查询处理的程序。包含了大量的数组操作、簇操作等等。
去下下来看看有没有刚好可以用到的地方。
呵呵。
